ddht/v5_1/alexandria/partials/proof.py

```diff
@@ -29,6 +36,84 @@ def __str__(self) -> str:
         return f"{display_path(self.path)}: {self.value.hex()}"
 
 
+@to_tuple
+def get_padding_elements(
+    start_index: int, num_padding_chunks: int, path_bit_length: int
+) -> Iterable[ProofElement]:
+    """
+    Get the padding elements for a proof.
+
+    By decomposing the number of chunks needed into the powers of two which
+    make up the number we can construct the minimal right hand sub-tree(s)
+    needed to pad the hash tree.
+    """
+    for power_of_two in decompose_into_powers_of_two(num_padding_chunks):
+        depth = power_of_two.bit_length() - 1
+        left_index = start_index + power_of_two
+        left_path = chunk_index_to_path(left_index, path_bit_length)
+        padding_hash_tree_root = ZERO_HASHES[depth]
+        yield ProofElement(left_path[: path_bit_length - depth], padding_hash_tree_root)
+
+
+class DataPartial:
+    """
+    A wrapper around a partial data proof which allows data retrieval by
+    indexing or slicing.
+
+    Raise `IndexError` if the requested data is not part of the proof.
+    """
```

Comment:
Even after reading the tests I'm still not sure what this class is for. You pass in a bunch of (path, data) tuples, and then you can read them back out by searching for them by path or path range?

Reply:
Yeah, a `Proof` houses *some* amount of the actual data (maybe all of it, maybe a subsection).  This abstracts that away so that you can just treat it like it's a bytestring and directly access the data via slice/index mechanics.  I'll beef up the docstring a bit.